spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/InferShape.scala

```diff
@@ -16,79 +16,119 @@
+
 trait InferShape {
 
+  private[bigdl] var usedAsSrc = false
+
+  private[bigdl] var usedAsDest = false
+
   private[bigdl] var _inputShapeValue: Shape = null
 
-  private[bigdl] var _outputShapeValue: Array[Shape] = Array[Shape]()
+  private[bigdl] var _outputShapeValue: Shape = null
 
   private[bigdl] def inputShapeValue: Shape = _inputShapeValue
 
-  private[bigdl] def outputShapeValue: Array[Shape] = _outputShapeValue
+  private[bigdl] def outputShapeValue: Shape = _outputShapeValue
 
   // scalastyle:off
   private[bigdl] def inputShapeValue_=(value: Shape): Unit = {
     _inputShapeValue = value
   }
 
-  private[bigdl] def outputShapeValue_=(value: Array[Shape]): Unit = {
+  private[bigdl] def outputShapeValue_=(value: Shape): Unit = {
     _outputShapeValue = value
   }
   // scalastyle:on
 
   /**
    * We suppose the first dim is batch
    */
-  private[bigdl] def getInputShape(): Shape = {
+  private[bigdl] final def getInputShape(): Shape = {
     _inputShapeValue
   }
 
   /**
-   * Get the outputshape by index.
-   * @param index start from 0
-   * @return
+   * We suppose the first dim is batch
    */
-  private[bigdl] def getOutputShapeFor(index: Int): Shape = {
-    _outputShapeValue(index)
+  private[bigdl] final def getOutputShape(): Shape = {
+    outputShapeValue
   }
 
-  /**
-   * We suppose the first dim is batch
-   */
-  private[bigdl] def getOutputShape(): Shape = {
-    if (_outputShapeValue.length > 1) {
-      throw new RuntimeException(
-        "There are multiple outputs for this layer. Please use getInputShapeFor instead")
-    }
-    outputShapeValue(0)
+  private[bigdl] def inferShape(calcInputShape: Shape): Shape = {
+    val outputShape = computeOutputShape(calcInputShape)
+    this.outputShapeValue = outputShape
+    this.inputShapeValue = calcInputShape
+    outputShape
   }
 
   /**
    * Execute building logic and return the outputShape for the given inputShape.
    * NB: the first dim of inputShape is batch
    */
   private[bigdl] def build(inputShape: Shape): Shape = {
-    val outputShape = computeOutputShape(inputShape)
-    this._outputShapeValue ++ Array(outputShape)
-    this._inputShapeValue = inputShape
-    isBuilt = true
-    outputShape
+    inferShape(inputShape)
   }
 
-  private[bigdl] var isBuilt: Boolean = false
-
+  private[bigdl] def isBuilt(): Boolean = outputShapeValue != null
 
-  private[bigdl] def isCompatibleWithKeras(): Boolean = true
+  private[bigdl] def isKerasStyle(): Boolean = false
 
-  private[bigdl] def isCompatibleWithTorch(): Boolean = true
+  private[bigdl] def allowRebuilt(): Boolean = false
 
   /**
    * We suppose the first dim is batch
    */
   private[bigdl] def computeOutputShape(inputShape: Shape): Shape = {
     throw new RuntimeException("Haven't been implemented yet. Do not use it with Keras Layer")
   }
+
+  private def ensureNotShared[T: ClassTag](modules : Seq[AbstractModule[_, _, T]]): Unit = {
```

Comment:
This is to for using this as dest, and modules as src - correct for `inputs()`, and not current for `Sequential.add()`